### ISSUE TITLE
motifcounter

### ISSUE THREAD

**Issue by wkopp** [Bioconductor/Contributions#258]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/wkopp/motifcounter

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by mtmorgan** [comment]
This looks like a really well-written package, congratulations!

##  Build commands

* `R-dev CMD build motifcounter`: Works

* `R-dev CMD INSTALL motifcounter_0.99.0.tar.gz`: Works

* Good clean build with no issues as shown [here][]. Good job!

[here]: http://bioconductor.org/spb_reports/motifcounter_buildreport_20170123075819.html)

## DESCRIPTION

* The `Description:` in the file doesn't contain enough information about the
  package. It should be like the abstract of a paper and inform users what the
  functionality is instead of just what methods are being written.

* Include ‘R (>= 3.0.0)’ or whatever is appropriate for your package in the
  ‘Depends’ field

## NAMESPACE

* Looks good.

## R

* Well written code.

* Recheck wording on some warning messages as they are not correctly
  stated, which might cause confusion. Pay attention to `stop()`,
  `warning()` formating -- currently these assume standard screen
  width of 80 characters; consider `paste(strwrap(.), collapse="\n")`

- Consider using S4 rather than S3 objects; the formal class and
  method definition provide a much more robust environment for
  software development, and the changes are not substantial, e.g.,
  
    ```
    background = list(
        station = dummy[[2]],
        trans = dummy[[3]],
        counts = dummy[[1]],
        order = order
    )
    class(background) = "Background"
    ```


  might become

    ```
    .Background <- setClass(
        "Background",
        slots = c(
            station = "numeric",
            trans="numeric",
            counts="numeric"
            order="integer"
        )
    )
    ```
    
  with use
  
      ```
      .Background(
          station = dummy[[2]],
          trans = dummy[[3]],
          counts = dummy[[1]],
          order = order
      )
      ```

  The validity check `backgroundValid()` could be implemented (no need
  to check for `length()` or `class()` as:
  
      ```
      setValidity("Background", function(object) {
          msg <- character(0)

          if (length(object@trans) != 4 ^ (object@order + 1)) {
              msg <- c(msg, paste(
                  "Inconsistent Background object."
                  "Use readBackground() to construct bg."
              ))
          if (length(msg)) msg else TRUE
      })
      ```

  This and the use of generic + methods means that many of the checks
  at the start of each function don't need to be performed.

- Avoid `1:n`-style iteration, which produces surprising results when
  `n == 0`; use `seq_len(n)` or `seq_along()` instead.

- For similar robustness reasons, use `tail(x, -1)` to drop the first
  element

    ```
    > ind =1
    > ind[2:length(ind)]
    [1] NA  1
    > tail(ind, -1)
    numeric(0)
    ```
    
- Use `is()` (S4) or `inherits()` (S3) to test class, rather than
  `class() == `, as in count_wrapper.R:155 `stopifnot(class(seqs) ==
  "DNAStringSet")`

## C

- Generally this looks really good.

- Use `R_alloc()` rather than `Calloc()` / `Free()` for memory that is
  used and released during a single function call. This avoids memory
  leaks on error, for instance in Rcombinatorial.c:47 allocates memory
  with Calloc, and returns on error at Rcombinatorial.c:87 without
  Free'ing, resulting in a memory leak.

- Consider using symbols `motifcounter_combinatorialDist` rather than
  character(1) `"motifcounter_combinatorialDist"` in `.C()`. This is
  (slightly) faster and more robust.

- Consider consistent indentation (e.g., 4 spaces) and other
  formatting conventions to increase code clarity.

## vignette

* Vignette looks great in terms of styling!

- Consider a hard-coded 'last modified' date, rather than or in
  addition to the system-generated build date `Sys.Date()`.

- 'Bioconductor' rather than 'bioconductor'.

- Consider concluding `sessionInfo()` evaluated code chunk for when
  vignette is viewd out of context.

## man

* Well written man pages

## tests, etc.

- Nice use of test and other development infrastructure.

**Comment by wkopp** [comment]
Thank you very much for your positive review! The above stated points sound reasonable. I'll work on them in the next couple days.

**Comment by wkopp** [comment]
Dear @mtmorgan,

I've addressed almost all of the considerations.

# DESCRIPTION
I have added a more comprehensive Description as well as an explicit Depends on
R >= 3.0. Note that this caused warnings during the last build. How do I address that? Or can I ignore that warning?

# R
- I went over all error and warning messages and corrected the misleading ones.
- I introduced the S4 Classes `Background` and `Overlap`. I did not introduce new Generics and methods for them, since they are merely used as containers that hold specifically structured data and the user isn't meant to interact with those classes. These objects are mostly used behind the scenes.
- I left the validity checks at the beginning of various functions for the following reason: `Background`-objects are supposed to be constructed only via `readBackground`. They should not be changed by the user. If a user still tries to change the `Background`-object (e.g. by assigning something else to the slots), which causes the object to be inconsistent, the validity checks would notify the user about that.
However, if you think that is overly cautious, I can also remove those validity checks entirely.

# C
- I have tried to substitute `R_alloc` for `Calloc` and removed the `Free`s. However, this leads to highly unstable behaviour of the software. I have used valgrind in order to find if I have introduced any obvious mistakes, but unfortunately, I was not able to resolve that issue. So I left that change out for now.
I have, however, created a 'use_R_alloc' git branch where I have left the use of `R_alloc` in one simple case (src/background.c) to demonstrate the issue. If you find any obvious mistake there, I'd be happy to address it.

- I've changed .C and .Call calls to invoke symbols instead of characters.

- I've run astyle over the *.[ch] files to make the formatting consistent.

# Vignette
- The 'last change' date was included hard-coded, 'bioconductor' was corrected to 'Bioconductor' and the `sessionInfo` was included at the end of the vignette.

**Comment by mtmorgan** [comment]
Feel free to ignore the warning about R version dependency.

**Comment by mtmorgan** [comment]
For S4 -- it is usual to introduce slot 'accessors' -- often plain-old-functions, not exported -- and use these throughout rather than to use direct slot access. This is the object-oriented principle of defining an interface separate from the implementation.

For C -- the code on your branch looks superficially correct. Is there an easy way to reproduce the problem that you report with respect to unusual memory problems?

**Comment by wkopp** [comment]
Regarding S4: So it would be ok to define accessor functions like this:
```
getOrder <-function(obj) {
   return(obj@order)
}
```

Regarding the R_alloc issue:
A reproduction of the error is given by travis-ci on a Linux system (Job #40):
https://travis-ci.org/wkopp/motifcounter/builds/210484505

Another option would be to do the following:
```
git clone https://github.com/wkopp/motifcounter.git
git checkout use_R_alloc
R CMD build motifcounter
R CMD check motifcounter_0.99.3.tar.gz
```
These commands do not finish successfully on my linux development environment and on the travis-ci linux system.

**Comment by wkopp** [comment]
Regarding the R_alloc issue, I think I found at least one issue (and possibly also the solutions). Let me try to fix this. I'll comment on this in more detail as soon as I definitely know what is going on, which will be in the next couple of days.

**Comment by wkopp** [comment]
Dear @mtmorgan,

- Regarding the S4 accessor function: I have added plain-old accessor functions to retrieve information from S4-object-slots.

- Regarding the `R_alloc` usage, I have found the issue. While, `Calloc` automatically initializes the allocated memory, `R_alloc` does not. I had assumed that they both would behave the same. memsetting of the transiently allocated memory solved the issue. In the current version, all `Calloc`s are now replaced with `R_alloc`.

Best,
Wolfgang
